model/glove/embedding.go

```diff
@@ -0,0 +1,98 @@
+// Copyright © 2017 Makoto Ito
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package glove
+
+import (
+	"gorgonia.org/tensor"
+
+	"github.com/ynqa/word-embedding/model"
+)
+
+// slice implements a single valued slice
+type slice int
+
+func (s slice) Start() int { return int(s) }
+func (s slice) End() int   { return int(s + 1) }
+func (s slice) Step() int  { return 1 }
+
+// Embedding represents a word embedding. It holds a Tensor, and preslices it for additional performance gains.
+type Embedding struct {
+	P []tensor.Tensor
+	Q []tensor.Tensor
+
+	gradP []tensor.Tensor
```

Comment:
To think about: is it necessary to keep a copy of the tensors of the gradients?

Reply:
Yes. GloVe uses AdaGrad as the optimizer, instead of SGD. AdaGrad also use all past iteration's gradients to decrease learning rate.
Here: https://en.wikipedia.org/wiki/Stochastic_gradient_descent#AdaGrad

Reply:
Yeah I'm familiar with Adagrad. I'm thinking maybe this shouldn't be put into the `Embedding` struct. Instead have something like a `Solver` struct

Reply:
ah I see, that's right. I'll try it.
(It's also good to select other optimizers, `AdaDelta`, `Adam`, and so on.)